kernel/init.c

```diff
@@ -219,18 +218,17 @@ static void bg_thread_main(void *unused1, void *unused2, void *unused3)
 	z_stack_adjust_initialized = 1;
 #endif
 	if (boot_delay > 0 && IS_ENABLED(CONFIG_MULTITHREADING)) {
-		printk("***** delaying boot " STRINGIFY(CONFIG_BOOT_DELAY)
-		       "ms (per build configuration) *****\n");
-		k_busy_wait(CONFIG_BOOT_DELAY * USEC_PER_MSEC);
+		printk("***** Delaying boot %u ms *****\n", boot_delay);
+		k_msleep(boot_delay);
```

Comment:
Shouldn't we use `k_busy_wait()` here like in the original version?

Reply:
With `k_msleep()` I could get the `printk()` message on UART before the boot_delay elapsed, with `k_busy_wait()` only after the delay elapsed and by then it is not useful anymore (me as a user was wondering what is happening).

Reply:
What platform?   That sounds like your logging is setup async, and doing a sleep allows a thread to run to do the output.  If so... that's not a bug, it's what you asked for, technically.  In general, printk() isn't synchronous any more (and yes, I hate this too).

Sleeping in the main thread before entry to main() is really pretty questionable anyway -- it allows all the threads to run **except** main, which is going to be very surprising to a typical user.

Reply:
> What platform?

It's a custom board based on STM32F412, but can also be seen on nrf52840-DK.

> That sounds like your logging is setup async, and doing a sleep allows a thread to run to do the output. If so... that's not a bug, it's what you asked for, technically. 

CONFIG_LOG_IMMEDIATE is indeed n, which is the default. Unsure whether I can switch it on in combination with OpenThread since that moves the logging into its context. Technically correct, still not exactly how I would have hoped

> In general, printk() isn't synchronous any more (and yes, I hate this too).

Was that an intentional change? Is there a way - that I am unaware of - to force certain prints to be immediate?

> Sleeping in the main thread before entry to main() is really pretty questionable anyway -- it allows all the threads to run **except** main, which is going to be very surprising to a typical user.

By that time of the boot process, are there any other threads than main?

Reply:
I would advice not using CONFIG_LOG_IMMEDIATE in any production system. I have been bit by this in networking where using the immediate option can slow things down and cause timeouts in the net stack, especially if one has various debugging options enabled.

There are networking TX & RX threads + other net related threads (depending on config) running before app main() called.

Reply:
Is there a printk_immediate() which I am unaware of? In cases where you want to make sure that it is immediately printed?

Reply:
how about using `printf()` and `k_busy_wait()` here? would that be acceptable?

Reply:
how about this as a solution for immediate printing:

```c
void immediate_printk(const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vprintk(fmt, ap);
	va_end(ap);
}
```

```c
	if (boot_delay > 0 && IS_ENABLED(CONFIG_MULTITHREADING)) {
		immediate_printk("***** Delaying boot %u ms *****\n",
			boot_delay);
		k_busy_wait(CONFIG_BOOT_DELAY * USEC_PER_MSEC);
	}
```

Reply:
Branch updated with this proposal. Works fine for us.